modules/objdetect/test/test_charucodetection.cpp

```diff
@@ -188,30 +179,31 @@ void CV_CharucoDetection::run(int) {
  */
 class CV_CharucoPoseEstimation : public cvtest::BaseTest {
     public:
-    CV_CharucoPoseEstimation();
+    CV_CharucoPoseEstimation(bool _legacyPattern) : legacyPattern(_legacyPattern) {}
 
     protected:
     void run(int);
-};
 
-
-CV_CharucoPoseEstimation::CV_CharucoPoseEstimation() {}
+    bool legacyPattern;
+};
 
 
 void CV_CharucoPoseEstimation::run(int) {
     int iter = 0;
     Mat cameraMatrix = Mat::eye(3, 3, CV_64FC1);
-    Size imgSize(500, 500);
+    Size imgSize(750, 750);
     aruco::DetectorParameters params;
     params.minDistanceToBorder = 3;
     aruco::CharucoBoard board(Size(4, 4), 0.03f, 0.015f, aruco::getPredefinedDictionary(aruco::DICT_6X6_250));
+    board.setLegacyPattern(legacyPattern);
     aruco::CharucoDetector detector(board, aruco::CharucoParameters(), params);
 
-    cameraMatrix.at<double>(0, 0) = cameraMatrix.at< double >(1, 1) = 650;
+    cameraMatrix.at<double>(0, 0) = cameraMatrix.at< double >(1, 1) = 1000;
     cameraMatrix.at<double>(0, 2) = imgSize.width / 2;
     cameraMatrix.at<double>(1, 2) = imgSize.height / 2;
 
     Mat distCoeffs(5, 1, CV_64FC1, Scalar::all(0));
+
     // for different perspectives
     for(double distance = 0.2; distance <= 0.3; distance += 0.1) {
```

Comment:
Good point!
We should not have `for` loops with floating point variables.

Correct code should look like this:

```.diff
-for(double distance = 0.2; distance <= 0.3; distance += 0.1) {
+for (int i = 0; i < 2; ++i) {
+    double distance = 0.2 + i * 0.1;
```

---

It is better to do that separately as there are many cases bring from `opencv_contrib`:

```
./modules/objdetect/test/test_arucodetection.cpp:250:    for(double distance = 0.1; distance < 0.7; distance += 0.2) {
./modules/objdetect/test/test_boarddetection.cpp:54:    for(double distance = 0.2; distance <= 0.4; distance += 0.15) {
./modules/objdetect/test/test_boarddetection.cpp:165:    for(double distance = 0.2; distance <= 0.4; distance += 0.2) {
./modules/objdetect/test/test_charucodetection.cpp:112:    for(double distance = 0.2; distance <= 0.4; distance += 0.2) {
./modules/objdetect/test/test_charucodetection.cpp:216:    for(double distance = 0.2; distance <= 0.3; distance += 0.1) {
./modules/objdetect/test/test_charucodetection.cpp:335:    for(double distance = 0.2; distance <= 0.3; distance += 0.1) {
```

/cc @AleksandrPanov @asmorkalov

Reply:
I was more thinking of just explicitly listing the values to be iterated over:
```diff
-for(double distance = 0.2; distance <= 0.3; distance += 0.1) {
+for(double distance : {0.2, 0.3}) {
```

Reply:
I created a separate PR for this: https://github.com/opencv/opencv/pull/23271. I think listing the values `0.2` and `0.3` explicitly is more straightforward than using a temporary value (`int i`), at least for these small `for` loops.

The other PR is only in draft because this PR here needs to be merged first, or there will be a conflict. Both PRs require the Charuco test pattern resolution to be increased.